### PR TITLE
DEPR: MultiIndex.to_hierarchical

### PR DESCRIPTION
**Pull request by KalyanGokhale**
xref #18262
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/tests/indexes/test_multi.py [hunk 1]
```diff
@@ -1704,6 +1704,11 @@ def test_to_hierarchical(self):
         tm.assert_index_equal(result, expected)
         assert result.names == index.names
 
+        # GH21613
+        # .to_hierarchical will be deprecated
+        with tm.assert_produces_warning(FutureWarning):
+            result = index.to_hierarchical(2)
```
Comment (jreback) [inline pandas/tests/indexes/test_multi.py]: you don't need the result =

File: pandas/tests/indexes/test_multi.py [hunk 2]
```diff
@@ -1704,6 +1704,11 @@ def test_to_hierarchical(self):
         tm.assert_index_equal(result, expected)
```
Comment (jreback) [inline pandas/tests/indexes/test_multi.py]: you need to catch all uses of to_hierarchical (or you get the warnings)

File: pandas/core/indexes/multi.py [hunk 3]
```diff
@@ -1182,6 +1182,8 @@ def to_frame(self, index=True):
 
     def to_hierarchical(self, n_repeat, n_shuffle=1):
         """
```
Comment (jreback) [inline pandas/core/indexes/multi.py]: to_hierarchical is listed elsewhere in this file in a doc-string

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexes/multi.py (hunk 3 above)
Comment (KalyanGokhale) [reply pandas/core/indexes/multi.py]: Thanks - it is mentioned as a part of enumeration of methods of `MultiIndex` - should it be removed?

File: pandas/tests/indexes/test_multi.py (hunk 2 above)
Comment (KalyanGokhale) [reply pandas/tests/indexes/test_multi.py]: Thanks - not sure if I understand. 
The only use of `to_hierarchical` was in `pandas/core/panel.py`- this has been substituted. 
This existing test is specifically for `to_hierarchical` - which can be removed once this method is removed in a future version - should these tests be removed as part of this PR?

File: pandas/tests/indexes/test_multi.py (hunk 1 above)
Comment (KalyanGokhale) [reply pandas/tests/indexes/test_multi.py]: Thanks - yes had missed it - will update together along with other comments

File: pandas/tests/indexes/test_multi.py (hunk 2 above)
Comment (KalyanGokhale) [reply pandas/tests/indexes/test_multi.py]: @jreback I think after reviewing changes in #18258 I probably get what is expected - please ignore the comment above - will update

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexes/multi.py (hunk 3 above)
Comment (KalyanGokhale) [reply pandas/core/indexes/multi.py]: Thanks - updated for all of the latest review comments

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks!

**Pull request merged**